web_client/components/body/analysis/element/feature-selection.jsx

```diff
@@ -0,0 +1,238 @@
+import React from 'react';
+import TextField from '../../../common/text-field.jsx';
+
+import { TIMES } from '../../../../constants';
+
+const addEntries = (currentEntries, newEntries) => {
+  let result = [].concat(currentEntries);
+  (
+    newEntries
+    .forEach((e) => {
+      if (result.indexOf(e) < 0) {
+        result.push(e);
+      }
+    })
+  );
+
+  return result;
+};
+
+class FeatureSelection extends React.Component {
+  render () {
+    let {
+      action,
+      mainAction,
+      onAction,
+      onStateChange,
+      state
+    } = this.props;
+
+    action = action || mainAction;
+    state = state || {};
+
+    let {
+      candidates,
+      currentText,
+      value: chosenSet,
+      extendedText,
+      buttonVisible
+    } = state;
+
+    candidates = candidates || [];
+    currentText = currentText || '';
+    chosenSet = chosenSet || [];
+    extendedText = extendedText || '';
+    buttonVisible = buttonVisible || false;
+
+    const handleNewEntries = () => {
+      onStateChange({
+        value: addEntries(
+          chosenSet,
+          (
+            [state.currentText].concat(
+              extendedText.split('\n')
+            )
+
+            .map((s) => (
+              s
+              .split(' ')
+              .filter(({ length }) => length)
+              .join(' ')
+            ))
+
+            .filter(({ length }) => length)
```

Comment:
This seems confusing. Is the destructuring argument meant to take advantage of the fact that strings and arrays have a static `length` property? I would change this to be a little bit less opaque:

```javascript
.filter(s => s.length > 0)
```